### PR TITLE
feat: Add OTP login APIs

### PR DESCRIPTION
**Pull request by howar31**
# Issue
Add OTP login APIs
- /v1/mail/send_otp
- /v3/auth/signin
- /v3/auth/activate

# Notice

# Dependency

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204934606026373

### CODE REVIEW

File: controllers/account.go [hunk 1]
```diff
@@ -382,3 +383,216 @@ func (mc *MembershipController) Onboarding(c *gin.Context) (int, gin.H, error) {
...
+			Email:         email,
+			ActivateToken: otpCode,
+			// expire time is 15 minute
+			ActExpTime: time.Now().Add(time.Duration(15) * time.Minute),
+		}
+
+		// try to find record by email in users table
+		matchedUser, err = mc.Storage.GetUserByEmail(email)
+		// the user record is not existed
+		if err != nil {
```
Comment (Aylie-Chou) [inline controllers/account.go:455]: Question: should we determine error case or matchedUser nil case here ?

File: controllers/account.go [hunk 2]
```diff
@@ -382,3 +383,216 @@ func (mc *MembershipController) Onboarding(c *gin.Context) (int, gin.H, error) {
...
+
+	var activatedString string
+	if user.Activated.Valid && !user.Activated.Time.IsZero() {
+		activatedString = user.Activated.Time.Format(time.RFC3339)
+	}
+
+	c.SetCookie("activated", activatedString, idTokenExpiration, defaultPath, defaultDomain, secure, true)
+	c.SetCookie("id_token", idToken, idTokenExpiration, defaultPath, defaultDomain, secure, true)
+
+	return http.StatusOK, gin.H{
```
[the remark this answers is not in the document]
Comment (howar31) [reply controllers/account.go:592]: As discussion in meeting, the redirect will be implemented by frontend.  Backend will set cookie header only.

File: controllers/account.go [hunk 3]
```diff
@@ -382,3 +383,216 @@ func (mc *MembershipController) Onboarding(c *gin.Context) (int, gin.H, error) {
...
+				Email: email,
+			},
+		}, errors.New(errMsg)
+	}
+
+	// check expire time
+	if ra.ActExpTime.Before(time.Now()) {
+		errMsg := "otp code expired"
+		return http.StatusForbidden, gin.H{
+			"status": "fail",
```
Comment (Aylie-Chou) [inline controllers/account.go:543]: Could we use `expired` as status value here to help frontend identify the fail cases 🙏

Comment (howar31) [reply controllers/account.go]: done

**Review by Aylie-Chou — CHANGES_REQUESTED** [review]
There's one comment left, please have a look ~
Also could we have test cases for the new added function & update `/v3/activate` route `403` error info in api doc in next PR (not blocking this feature, but also important) 🙏

File: controllers/account.go (hunk 1 above)
Comment (howar31) [reply controllers/account.go:455]: According to other similar functions (L110, L133), we do not need to determine error case or matchedUser nil case here.

Comment (Aylie-Chou) [reply controllers/account.go:455]: If `matchedUser` and `err` are both nil, current code would still mark the user existed.
Is that the expected result ?

**Review by Aylie-Chou — COMMENTED** [review]
Please have a look at the new added comment.

File: controllers/account.go (hunk 1 above)
Comment (howar31) [reply controllers/account.go:455]: I tried the `GetUserByEmail()`.  If `matchedUser` is nil (user not found), the `err` will return `record not found` error.  Normally they won't both be nil.

Comment (Aylie-Chou) [reply controllers/account.go:455]: Got it !
Thanks for the checking.

**Review by Aylie-Chou — APPROVED** [review]
LGTM

**Pull request merged**